src/core/experimental/refcount.d

```diff
@@ -0,0 +1,545 @@
+module core.experimental.refcount;
+
+private struct StatsAllocator
+{
+    version(CoreUnittest) size_t bytesUsed;
+
+    @trusted @nogc nothrow pure
+    void* allocate(size_t bytes) shared
+    {
+        import core.memory : pureMalloc;
+        if (!bytes) return null;
+
+        auto p = pureMalloc(bytes);
+        if (p is null) return null;
+        enum alignment = size_t.sizeof;
+        assert(cast(size_t) p % alignment == 0);
+
+        version (CoreUnittest)
+        {
+            static if (is(typeof(this) == shared))
+            {
+                import core.atomic : atomicOp;
+                atomicOp!"+="(bytesUsed, bytes);
+            }
+            else
+            {
+                bytesUsed += bytes;
+            }
+        }
+        return p;
+    }
+
+    @system @nogc nothrow pure
+    bool deallocate(void[] b) shared
+    {
+        import core.memory : pureFree;
+        assert(b !is null);
+
+        version (CoreUnittest)
+        {
+            static if (is(typeof(this) == shared))
+            {
+                import core.atomic : atomicOp;
+                assert(atomicOp!">="(bytesUsed, b.length));
+                atomicOp!"-="(bytesUsed, b.length);
+            }
+            else
+            {
+                assert(bytesUsed >= b.length);
+                bytesUsed -= b.length;
+            }
+        }
+        pureFree(b.ptr);
+        return true;
+    }
+
+    static shared StatsAllocator instance;
+}
+
+version (CoreUnittest)
+{
+    private shared StatsAllocator allocator;
+
+    private @nogc nothrow pure @trusted
+    void* pureAllocate(size_t n)
+    {
+        return (cast(void* function(size_t) @nogc nothrow pure)(&_allocate))(n);
+    }
+
+    private @nogc nothrow @safe
+    void* _allocate(size_t n)
+    {
+        return allocator.allocate(n);
+    }
+
+    private @nogc nothrow pure
+    void* pureDeallocate(T)(T[] b)
+    {
+        return (cast(void* function(T[]) @nogc nothrow pure)(&_deallocate!(T)))(b);
+    }
+
+    private @nogc nothrow
+    void* _deallocate(T)(T[] b)
+    {
+        allocator.deallocate(b);
+        return null;
+    }
+}
+
+struct __RefCount
```

Comment:
Why the `__` prefix?

Reply:
`__` is very problematic is it's reserved by the compiler and can lead to strange errors. `_` should be used for library symbols.

Reply:
The prefix is to make the user aware that there is some `magic` going on under the hood in order to make it work.

Reply:
If there is no specific compiler support, there is no magic. Please note that there are subtle differences in how the compiler treats these symbols, e.g. https://github.com/dlang/dmd/blob/master/src/dmd/traits.d#L1395

Reply:
@edi33416 please use double underscore. This is a special/magic type that _is_ meant to be recognized by the compiler as special, even if today it happens to not be. Thanks.